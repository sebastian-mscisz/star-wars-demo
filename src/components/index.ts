export * from './character-details';
export * from './characters';
export * from './characters-card';
export * from './error-component';
export * from './items-with-pagination';
export * from './location-details';
export * from './locations';
export * from './locations-card';
export * from './marginal-components';
export * from './pagination-buttons';
export * from './spinner';
export * from './vehicle-details';
export * from './vehicles';
export * from './vehicles-card';
