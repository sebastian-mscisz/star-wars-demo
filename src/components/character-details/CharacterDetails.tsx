import React, { useEffect } from 'react';
import { CharacterDetailsUi } from './ui/CharacterDetailsUi';
import { withHandlers } from 'src/hoc';
import { useCharacterDetails, useLocations, useSpecies, useVehicles } from 'src/hooks';
import { useNavigate } from 'react-router';
import { notFoundPagePath } from 'src/common';

const CharacterDetailsWithHandlers = withHandlers(CharacterDetailsUi);

type CharacterDetailsProps = {
    characterId: string;
};

export const CharacterDetails: React.FC<CharacterDetailsProps> = ({ characterId }) => {
    const navigate = useNavigate();
    const { data: character, isLoading, loadingError, hasReceivedResponse } = useCharacterDetails(characterId);
    const { data: locations, isLoading: isLoadingLocations, loadingError: loadingErrorLocations } = useLocations();
    const { data: vehicles, isLoading: isLoadingVehicles, loadingError: loadingErrorVehicles } = useVehicles();
    const { data: species, isLoading: isLoadingSpecies, loadingError: loadingErrorSpecies } = useSpecies();

    useEffect(() => {
        if (hasReceivedResponse && character === undefined) navigate(notFoundPagePath());
    }, [hasReceivedResponse, character, navigate]);

    return (
        <CharacterDetailsWithHandlers
            isLoading={isLoading || isLoadingLocations || isLoadingSpecies || isLoadingVehicles}
            character={character}
            characterId={characterId}
            locations={locations}
            vehicles={vehicles}
            species={species}
            hasError={!!loadingError || !!loadingErrorLocations || !!loadingErrorSpecies || !!loadingErrorVehicles}
        />
    );
};
