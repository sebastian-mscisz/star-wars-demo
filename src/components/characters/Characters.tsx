import { withHandlers } from 'src/hoc';
import { CharactersUi } from './ui/CharactersUi';
import { useCharacters } from 'src/hooks';

const CharactersWithHandlers = withHandlers(CharactersUi);

export const Characters: React.FC = () => {
    const { data: characters, isLoading, loadingError } = useCharacters();

    return <CharactersWithHandlers isLoading={isLoading} characters={characters} hasError={!!loadingError} />;
};
