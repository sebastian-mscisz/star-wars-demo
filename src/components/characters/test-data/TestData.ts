import { Characters, Character, Species } from 'src/common';

export const TestCharacter: Character = {
    name: 'Character1',
    url: 'https://swapi.info/api/people/1',
    homeworld: 'https://swapi.info/api/planets/1',
    species: ['species/1'],
    vehicles: ['https://swapi.info/api/vehicles/1'],
};

export const TestCharacters: Characters = [
    TestCharacter,
    {
        name: 'Character2',
        url: 'https://swapi.info/api/people/2',
        vehicles: ['https://swapi.info/api/vehicles/1', 'https://swapi.info/api/vehicles/2'],
        homeworld: 'https://swapi.info/api/planets/2',
        species: ['species/2'],
    },
    {
        name: 'Character3',
        url: 'https://swapi.info/api/people/3',
        vehicles: [],
        homeworld: 'https://swapi.info/api/planets/3',
        species: ['species/3'],
    },
    {
        name: 'Character4',
        url: 'https://swapi.info/api/people/4',
        vehicles: ['https://swapi.info/api/vehicles/3'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/4'],
    },
    {
        name: 'Character5',
        url: 'https://swapi.info/api/people/5',
        vehicles: ['https://swapi.info/api/vehicles/4', 'https://swapi.info/api/vehicles/5'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/5'],
    },
];

export const TestSpecies: Species = [
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species3',
        url: 'species/3',
    },
    {
        name: 'Species4',
        url: 'species/4',
    },
    {
        name: 'Species5',
        url: 'species/5',
    },
];
