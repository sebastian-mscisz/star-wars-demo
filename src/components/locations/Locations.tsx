import { withHandlers } from 'src/hoc';
import { LocationsUi } from './ui/LocationsUi';
import { useLocations } from 'src/hooks';

const LocationsWithHandlers = withHandlers(LocationsUi);

export const Locations: React.FC = () => {
    const { data: locations, isLoading, loadingError } = useLocations();

    return <LocationsWithHandlers isLoading={isLoading} locations={locations} hasError={!!loadingError} />;
};
