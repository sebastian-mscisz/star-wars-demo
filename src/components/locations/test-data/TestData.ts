import { Locations, Location } from 'src/common';

export const TestLocation: Location = {
    name: 'Location1',
    url: 'https://swapi.info/api/planets/1',
    residents: ['https://swapi.info/api/people/1'],
    population: '1000',
};

export const TestLocations: Locations = [
    TestLocation,
    {
        name: 'Location2',
        url: 'https://swapi.info/api/planets/2',
        residents: ['https://swapi.info/api/people/2'],
        population: '5343',
    },
    {
        name: 'Location3',
        url: 'https://swapi.info/api/planets/3',
        residents: ['https://swapi.info/api/people/3'],
        population: '13333',
    },
    {
        name: 'Location4',
        url: 'https://swapi.info/api/planets/4',
        residents: ['https://swapi.info/api/people/4', 'https://swapi.info/api/people/5'],
        population: '9999',
    },
    {
        name: 'Location5',
        url: 'https://swapi.info/api/planets/5',
        residents: [],
        population: '999',
    },
];
