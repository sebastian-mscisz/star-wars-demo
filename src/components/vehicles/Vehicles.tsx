import { withHandlers } from 'src/hoc';
import { VehiclesUi } from './ui/VehiclesUi';
import { useVehicles } from 'src/hooks';

const VehiclesWithHandlers = withHandlers(VehiclesUi);

export const Vehicles: React.FC = () => {
    const { data: vehicles, isLoading, loadingError } = useVehicles();

    return <VehiclesWithHandlers isLoading={isLoading} vehicles={vehicles} hasError={!!loadingError} />;
};
