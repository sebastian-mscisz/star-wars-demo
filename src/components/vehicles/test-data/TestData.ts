import { Vehicles, Vehicle } from 'src/common';

export const TestVehicle: Vehicle = {
    name: 'Vehicle1',
    url: 'https://swapi.info/api/vehicles/1',
    vehicle_class: 'Air speeder',
    pilots: ['character/1', 'character/2'],
};

export const TestVehicles: Vehicles = [
    TestVehicle,
    {
        name: 'Vehicle2',
        url: 'https://swapi.info/api/vehicles/2',
        pilots: ['character/2'],
        vehicle_class: 'Land speeder',
    },
    {
        name: 'Vehicle3',
        url: 'https://swapi.info/api/vehicles/3',
        pilots: ['character/4'],
        vehicle_class: 'Water speeder',
    },
    {
        name: 'Vehicle4',
        url: 'https://swapi.info/api/vehicles/4',
        pilots: ['character/5'],
        vehicle_class: 'Fire speeder',
    },
    {
        name: 'Vehicle5',
        url: 'https://swapi.info/api/vehicles/5',
        pilots: ['character/5'],
        vehicle_class: 'Air speeder',
    },
];
