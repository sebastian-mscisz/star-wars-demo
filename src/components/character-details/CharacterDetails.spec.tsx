import { BrowserRouter } from 'react-router-dom';
import { CharacterDetails } from './CharacterDetails';
import { screen, render } from '@testing-library/react';
import MockAdapter from 'axios-mock-adapter';
import { DataApi } from 'src/xhr';
import { getCharactersEndpoint, getLocationsEndpoint, getSpeciesEndpoint, getVehiclesEndpoint } from 'src/common';
import { Provider } from 'react-redux';
import { createStore } from '../../store';
import { TestCharacters, TestSpecies } from '../characters/test-data';
import { TestVehicles } from '../vehicles/test-data';
import { TestLocations } from '../locations/test-data';

describe('CharacterDetails', () => {
    let dataApiMock: MockAdapter;

    beforeEach(() => {
        dataApiMock = new MockAdapter(DataApi);
    });

    it('should render', async () => {
        dataApiMock.onGet(getCharactersEndpoint()).reply(200, TestCharacters);
        dataApiMock.onGet(getVehiclesEndpoint()).reply(200, TestVehicles);
        dataApiMock.onGet(getLocationsEndpoint()).reply(200, TestLocations);
        dataApiMock.onGet(getSpeciesEndpoint()).reply(200, TestSpecies);

        const { asFragment } = render(
            <Provider store={createStore()}>
                <CharacterDetails characterId="1" />
            </Provider>,
            { wrapper: BrowserRouter },
        );

        await screen.findByText('Character1');

        expect(asFragment()).toMatchSnapshot();
    });

    it('should error component on characters error response', async () => {
        dataApiMock.onGet(getCharactersEndpoint()).reply(500);
        dataApiMock.onGet(getVehiclesEndpoint()).reply(200, TestVehicles);
        dataApiMock.onGet(getLocationsEndpoint()).reply(200, TestLocations);
        dataApiMock.onGet(getSpeciesEndpoint()).reply(200, TestSpecies);

        render(
            <Provider store={createStore()}>
                <CharacterDetails characterId="2" />
            </Provider>,
            { wrapper: BrowserRouter },
        );

        await screen.findByText('500 No connection. Please try again later.');
    });

    it('should error component on vehicles error response', async () => {
        dataApiMock.onGet(getCharactersEndpoint()).reply(200, TestCharacters);
        dataApiMock.onGet(getVehiclesEndpoint()).reply(500);
        dataApiMock.onGet(getLocationsEndpoint()).reply(200, TestLocations);
        dataApiMock.onGet(getSpeciesEndpoint()).reply(200, TestSpecies);

        render(
            <Provider store={createStore()}>
                <CharacterDetails characterId="3" />
            </Provider>,
            { wrapper: BrowserRouter },
        );

        await screen.findByText('500 No connection. Please try again later.');
    });

    it('should error component on locations error response', async () => {
        dataApiMock.onGet(getCharactersEndpoint()).reply(200, TestCharacters);
        dataApiMock.onGet(getVehiclesEndpoint()).reply(200, TestVehicles);
        dataApiMock.onGet(getLocationsEndpoint()).reply(500);
        dataApiMock.onGet(getSpeciesEndpoint()).reply(200, TestSpecies);

        render(
            <Provider store={createStore()}>
                <CharacterDetails characterId="3" />
            </Provider>,
            { wrapper: BrowserRouter },
        );

        await screen.findByText('500 No connection. Please try again later.');
    });

    it('should error component on species error response', async () => {
        dataApiMock.onGet(getCharactersEndpoint()).reply(200, TestCharacters);
        dataApiMock.onGet(getVehiclesEndpoint()).reply(200, TestVehicles);
        dataApiMock.onGet(getLocationsEndpoint()).reply(200, TestLocations);
        dataApiMock.onGet(getSpeciesEndpoint()).reply(500);

        render(
            <Provider store={createStore()}>
                <CharacterDetails characterId="3" />
            </Provider>,
            { wrapper: BrowserRouter },
        );

        await screen.findByText('500 No connection. Please try again later.');
    });
});
