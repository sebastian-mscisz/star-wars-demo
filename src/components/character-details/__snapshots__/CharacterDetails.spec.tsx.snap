// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetails should render 1`] = `
<DocumentFragment>
  <div
    class="sw-character-detail"
  >
    <div
      class="sw-character-detail__name"
    >
      <div
        class="sw-character-detail__name__image-wrapper"
      >
        <div
          class="sw-character-detail__name__image-wrapper__display-spinner"
        >
          <div
            class="sw-spinner"
          >
            <span
              style="display: inherit; position: relative; width: 25px; height: 5px; overflow: hidden; background-color: black; background-clip: padding-box; border-radius: 5px;"
            >
              <span
                style="position: absolute; height: 5px; overflow: hidden; background-color: rgb(0, 0, 0); background-clip: padding-box; display: block; border-radius: 2px; will-change: left, right; animation-fill-mode: forwards; animation: react-spinners-BarLoader-long Infinitys  cubic-bezier(0.65, 0.815, 0.735, 0.395) infinite;"
              />
              <span
                style="position: absolute; height: 5px; overflow: hidden; background-color: rgb(0, 0, 0); background-clip: padding-box; display: block; border-radius: 2px; will-change: left, right; animation-fill-mode: forwards; animation: react-spinners-BarLoader-short Infinitys Infinitys cubic-bezier(0.165, 0.84, 0.44, 1) infinite;"
              />
            </span>
            <span
              style="display: inherit; position: relative; width: 100px; height: 5px; overflow: hidden; background-color: rgba(0, 0, 0, 0.2); background-clip: padding-box; border-radius: 5px;"
            >
              <span
                style="position: absolute; height: 5px; overflow: hidden; background-color: rgb(0, 0, 0); background-clip: padding-box; display: block; border-radius: 2px; will-change: left, right; animation-fill-mode: forwards; animation: react-spinners-BarLoader-long 2.1s  cubic-bezier(0.65, 0.815, 0.735, 0.395) infinite;"
              />
              <span
                style="position: absolute; height: 5px; overflow: hidden; background-color: rgb(0, 0, 0); background-clip: padding-box; display: block; border-radius: 2px; will-change: left, right; animation-fill-mode: forwards; animation: react-spinners-BarLoader-short 2.1s 1.15s cubic-bezier(0.165, 0.84, 0.44, 1) infinite;"
              />
            </span>
          </div>
        </div>
        <img
          alt="Star wars character"
          class="sw-character-detail__name__image-wrapper__hide_img"
          src="./characters/1.jpg"
        />
      </div>
      <h3>
        Character1
      </h3>
    </div>
    <div
      class="sw-character-detail__details"
    >
      <p>
        <strong>
          Origin world:
        </strong>
         
        <a
          class="sw-character-detail__details__link"
          href="/location/1"
        >
          Location1
        </a>
      </p>
      <hr />
      <p>
        <strong>
          Race:
        </strong>
         Unknown race
      </p>
      <hr />
      <p>
        <strong>
          Used vehicles:
        </strong>
         
        <a
          class="sw-character-detail__details__link"
          href="/vehicle/1"
        >
          Vehicle1
        </a>
      </p>
    </div>
  </div>
  <div
    class="sw-pagination-section sw-pagination-section--center"
    data-testid="pagination-section"
  >
    <div
      class="sw-pagination-button "
      data-testid="left-pagination-button"
    >
      ←
    </div>
    <div
      class="sw-pagination-button "
      data-testid="right-pagination-button"
    >
      →
    </div>
  </div>
</DocumentFragment>
`;
